File: android/src/main/java/com/stallion/networkmanager/StallionStageManager.java
package com.stallion.networkmanager;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;
import com.stallion.events.StallionEventConstants;
import com.stallion.events.StallionEventManager;
import com.stallion.storage.StallionConfig;
import com.stallion.storage.StallionConfigConstants;
import com.stallion.storage.StallionMetaConstants;
import com.stallion.storage.StallionStateManager;

import org.json.JSONObject;

public class StallionStageManager {
  public static void downloadStageBundle(ReadableMap bundleInfo, Promise promise) {
    StallionStateManager stallionStateManager = StallionStateManager.getInstance();
    String receivedDownloadUrl = bundleInfo.getString("url");
    String receivedHash = bundleInfo.getString("hash");
    if(
      receivedDownloadUrl != null
        && !receivedDownloadUrl.isEmpty()
        && receivedHash != null
        && !receivedHash.isEmpty()
    ) {
      StallionConfig config = stallionStateManager.getStallionConfig();
      String downloadPath = config.getFilesDirectory()
        + StallionConfigConstants.STAGE_DIRECTORY
        + StallionConfigConstants.TEMP_FOLDER_SLOT;
      long alreadyDownloaded = StallionDownloadCacheManager.getDownloadCache(config, receivedDownloadUrl, downloadPath);

      emitDownloadStartedStage(receivedHash, alreadyDownloaded > 0);

      StallionFileDownloader.downloadBundle(
        receivedDownloadUrl,
        downloadPath,
        alreadyDownloaded,
        new StallionDownloadCallback() {
          @Override
          public void onReject(String prefix, String error) {
            promise.reject(prefix, error);
            emitDownloadErrorStage(receivedHash, error);
          }

          @Override
          public void onSuccess(String successPayload) {
            stallionStateManager.stallionMeta.setCurrentStageSlot(StallionMetaConstants.SlotStates.NEW_SLOT);
            stallionStateManager.stallionMeta.setStageTempHash(receivedHash);
            stallionStateManager.syncStallionMeta();
            emitDownloadSuccessStage(receivedHash);
            StallionDownloadCacheManager.deleteDownloadCache(downloadPath);
            promise.resolve(successPayload);
          }

          @Override
          public void onProgress(double downloadFraction) {
            emitDownloadProgressStage(receivedHash, downloadFraction);
          }
        }
      );
    }
  }

  private static void emitDownloadSuccessStage(String releaseHash) {
    JSONObject successPayload = new JSONObject();
    try {
      successPayload.put("releaseHash", releaseHash);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      StallionEventConstants.NativeStageEventTypes.DOWNLOAD_COMPLETE_STAGE.toString(),
      successPayload
    );
  }

  private static void emitDownloadProgressStage(String releaseHash, double newProgress) {
    JSONObject successPayload = new JSONObject();
    try {
      successPayload.put("releaseHash", releaseHash);
      successPayload.put("progress", String.valueOf(newProgress));
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEventWithoutCaching(
      StallionEventConstants.NativeStageEventTypes.DOWNLOAD_PROGRESS_STAGE.toString(),
      successPayload
    );
  }

  private static void emitDownloadStartedStage(String releaseHash, Boolean isResume) {
    JSONObject startedPayload = new JSONObject();
    try {
      startedPayload.put("releaseHash", releaseHash);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      isResume ?
        StallionEventConstants.NativeStageEventTypes.DOWNLOAD_RESUME_STAGE.toString()
        : StallionEventConstants.NativeStageEventTypes.DOWNLOAD_STARTED_STAGE.toString(),
      startedPayload
    );
  }

  private static void emitDownloadErrorStage(String releaseHash, String error) {
    JSONObject errorPayload = new JSONObject();
    try {
      errorPayload.put("releaseHash", releaseHash);
      errorPayload.put("meta", error);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      StallionEventConstants.NativeStageEventTypes.DOWNLOAD_ERROR_STAGE.toString(),
      errorPayload
    );
  }
}

File: ios/main/StallionStageManager.swift
//
//  StallionStageManager.swift
//  react-native-stallion
//
//  Created by Jasbir Singh Shergill on 29/01/25.
//

import Foundation

@objc(StallionStageManager)
class StallionStageManager: NSObject {
    
    @objc static func downloadStageBundle(bundleInfo: NSDictionary, promise: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        // Validate input parameters
        guard let receivedDownloadUrl = bundleInfo["url"] as? String,
              !receivedDownloadUrl.isEmpty,
              let receivedHash = bundleInfo["hash"] as? String,
              !receivedHash.isEmpty else {
            rejecter("INVALID_INPUT", "Invalid or missing download URL or hash.", nil)
            return
        }
        
        // Ensure StallionStateManager instance exists
        guard let stallionStateManager = StallionStateManager.sharedInstance() else {
            rejecter("STATE_MANAGER_ERROR", "Unable to fetch StallionStateManager instance.", nil)
            return
        }
        
        // Construct the download path
        let downloadPath = (stallionStateManager.stallionConfig.filesDirectory ?? "") + "/" + StallionConstants.STAGE_DIRECTORY + "/" + (StallionConstants.TEMP_FOLDER_SLOT);
        
        // Capture resolve/reject methods to avoid escaping issues
        let resolveClosure: (Any?) -> Void = { result in
            DispatchQueue.main.async {
                promise(result)
            }
        }
        
        let rejectClosure: (String?, String, Error?) -> Void = { code, message, error in
            DispatchQueue.main.async {
                rejecter(code, message, error)
            }
        }
        
        emitDownloadStartedStage(releaseHash: receivedHash)
      
        // Start bundle download
        StallionFileDownloader().downloadBundle(
            url: URL(string: receivedDownloadUrl)!,
            downloadDirectory: downloadPath,
            onProgress: { progress in
                emitDownloadProgressStage(releaseHash: receivedHash, progress: progress)
            },
            resolve: { _ in
                stallionStateManager.stallionMeta.currentStageSlot = SlotStates.newSlot
                stallionStateManager.stallionMeta.stageTempHash = receivedHash
                stallionStateManager.syncStallionMeta()
                emitDownloadSuccessStage(releaseHash: receivedHash)
                resolveClosure("Stage download success")
            },
            reject: { code, prefix, error in
                let errorMessage = "\(prefix ?? "") \(error?.localizedDescription ?? "Unknown error")"
                emitDownloadErrorStage(releaseHash: receivedHash, error: errorMessage)
                rejectClosure(code, errorMessage, error)
            }
        )
    }
    
    // MARK: - Event Emitters
  
  private static func emitDownloadErrorStage(releaseHash: String, error: String) {
      let errorPayload: NSDictionary = [
          "releaseHash": releaseHash,
          "meta": error
      ]
    Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesStage.DOWNLOAD_ERROR_STAGE,
                           eventBody: errorPayload,
                           shouldCache: true
    )
  }
    
    private static func emitDownloadSuccessStage(releaseHash: String) {
      let successPayload: NSDictionary = ["releaseHash": releaseHash]
      Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesStage.DOWNLOAD_COMPLETE_STAGE,eventBody: successPayload,
                             shouldCache: true
      )
    }
    
    private static func emitDownloadProgressStage(releaseHash: String, progress: Float) {
        let progressPayload: NSDictionary = [
            "releaseHash": releaseHash,
            "progress": "\(progress)"
        ]
      Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesStage.DOWNLOAD_PROGRESS_STAGE, eventBody: progressPayload,
                             shouldCache: false
      )
    }
  private static func emitDownloadStartedStage(releaseHash: String) {
      let startedPayload: NSDictionary = ["releaseHash": releaseHash]
    Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesStage.DOWNLOAD_STARTED_STAGE,
                           eventBody: startedPayload,
                           shouldCache: true
    )
  }
}
